### PR TITLE
sys: shell: ping6: allow specification of target interface

### PR DESCRIPTION
**Pull request by kaspar030**
This PR changes the ping6 shell command to accept (and if needed, require) an interface specifier.

```
> ping6 fe80::1%6
```

If destination address is link local and only one interface present, use that interface.
If destination address is link local and multiple interfaces present, require interface specifier.
Otherwise, use specifier if present.

It also adds, if an interface specifier is present, the necessary netif header.

*edit* depends on #4753.

### CODE REVIEW

**Comment by kaspar030** [comment]
- hot-fixed corner case and amended

File: sys/shell/commands/sc_icmpv6_echo.c [hunk 1]
```diff
@@ -38,9 +38,10 @@ static char ipv6_str[IPV6_ADDR_MAX_STR_LEN];
 
 static void usage(char **argv)
 {
-    printf("%s [<count>] <ipv6 addr> [<payload_len>] [<delay in ms>] [<stats interval>]\n", argv[0]);
+    printf("%s [<count>] <ipv6 addr[%%<device>]> [<interface>] [<payload_len>] [<delay in ms>] [<stats interval>]\n", argv[0]);
```
Comment (authmillenon) [inline sys/shell/commands/sc_icmpv6_echo.c]: What's the difference between device any interface here?

Shouldn't the `[%%<device>]` go outside the variable defining `< >` of `<ipv6 addr>`?

File: sys/shell/commands/sc_icmpv6_echo.c [hunk 2]
```diff
@@ -148,6 +149,7 @@ int _icmpv6_ping(int argc, char **argv)
     uint32_t delay = 1 * SEC_IN_MS;
     char *addr_str;
     ipv6_addr_t addr;
+    kernel_pid_t src_iface = KERNEL_PID_UNDEF;
```
Comment (authmillenon) [inline sys/shell/commands/sc_icmpv6_echo.c]: You need to re-set this in 192 anyways, so why bother initializing?

File: sys/shell/commands/sc_icmpv6_echo.c (hunk 1 above)
Comment (kaspar030) [inline sys/shell/commands/sc_icmpv6_echo.c]: uh, I had an extra interface parameter before I decided to go the split route. fixed (and now it's "interface" everywhere)

File: sys/shell/commands/sc_icmpv6_echo.c (hunk 2 above)
Comment (kaspar030) [inline sys/shell/commands/sc_icmpv6_echo.c]: ypu, thx

**Comment by authmillenon** [comment]
Tested with `gnrc_border_router` and `gnrc_networking`. ACK and go when squashed and Travis is happy.

**Comment by authmillenon** [comment]
(oh and #4753 was merged)

**Comment by cgundogan** [comment]
rebase please

**Comment by cgundogan** [comment]
and squash while at it

**Comment by kaspar030** [comment]
- rebased, squashed

File: sys/shell/commands/sc_icmpv6_echo.c [hunk 3]
```diff
@@ -38,9 +38,10 @@ static char ipv6_str[IPV6_ADDR_MAX_STR_LEN];
 
 static void usage(char **argv)
 {
-    printf("%s [<count>] <ipv6 addr> [<payload_len>] [<delay in ms>] [<stats interval>]\n", argv[0]);
+    printf("%s [<count>] <ipv6 addr>[%%<interface>] [<payload_len>] [<delay in ms>] [<stats interval>]\n", argv[0]);
     puts("defaults:");
     puts("    count = 3");
+    puts("    interface = only interface if only one present, only needed for link-local addresses");
```
Comment (cgundogan) [inline sys/shell/commands/sc_icmpv6_echo.c]: so many onlys!

File: sys/shell/commands/sc_icmpv6_echo.c [hunk 4]
```diff
@@ -229,6 +262,19 @@ int _icmpv6_ping(int argc, char **argv)
             continue;
         }
 
+        if (src_iface != KERNEL_PID_UNDEF) {
+            pkt = gnrc_pktbuf_add(pkt, NULL,
+                    sizeof(gnrc_netif_hdr_t),
```
Comment (cgundogan) [inline sys/shell/commands/sc_icmpv6_echo.c]: indentation is off

File: sys/shell/commands/sc_icmpv6_echo.c (hunk 3 above)
Comment (authmillenon) [inline sys/shell/commands/sc_icmpv6_echo.c]: Only if you only-intolerant ;-P

**Comment by kaspar030** [comment]
- addressed comments & immediately squashed

File: sys/shell/commands/sc_icmpv6_echo.c [hunk 5]
```diff
@@ -212,7 +245,7 @@ int _icmpv6_ping(int argc, char **argv)
         uint32_t start, stop, timeout = 1 * SEC_IN_USEC;
 
         pkt = gnrc_icmpv6_echo_build(ICMPV6_ECHO_REQ, id, ++max_seq_expected,
-                                     NULL, payload_len);
+                NULL, payload_len);
```
Comment (authmillenon) [inline sys/shell/commands/sc_icmpv6_echo.c]: Indentation.

File: sys/shell/commands/sc_icmpv6_echo.c [hunk 6]
```diff
@@ -222,13 +255,25 @@ int _icmpv6_ping(int argc, char **argv)
         _set_payload(pkt->data, payload_len);
 
         pkt = gnrc_netreg_hdr_build(GNRC_NETTYPE_IPV6, pkt, NULL, 0, addr.u8,
-                                    sizeof(ipv6_addr_t));
+                sizeof(ipv6_addr_t));
```
Comment (authmillenon) [inline sys/shell/commands/sc_icmpv6_echo.c]: Indentation.

File: sys/shell/commands/sc_icmpv6_echo.c [hunk 7]
```diff
@@ -222,13 +255,25 @@ int _icmpv6_ping(int argc, char **argv)
...
 
         if (pkt == NULL) {
             puts("error: packet buffer full");
             continue;
         }
 
+        if (src_iface != KERNEL_PID_UNDEF) {
+            pkt = gnrc_pktbuf_add(pkt, NULL, sizeof(gnrc_netif_hdr_t),
```
Comment (authmillenon) [inline sys/shell/commands/sc_icmpv6_echo.c]: Indentation.

**Comment by kaspar030** [comment]
What are our indentation rules / where can I find them? Up to now, I trusted my vim indentation, but you seem not to. :)

**Comment by authmillenon** [comment]
I trust the uncrustify config more then the vim config (but my vim config is doing this too). It's supposed to be as it was before for the indentations you changed:

```C
            pkt = gnrc_pktbuf_add(pkt, NULL, sizeof(gnrc_netif_hdr_t),
                                  GNRC_NETTYPE_NETIF);
```

**Comment by authmillenon** [comment]
`set cino=(0` is the configuration you are searching for ;-)

**Comment by kaspar030** [comment]
Ok. I think I've fixed the indentation.

**Comment by authmillenon** [comment]
ACK and go, when Travis is happy.

**Pull request merged**